### ISSUE TITLE
Develop a check for undefined behaviour

### ISSUE THREAD

**Issue by lucasccordeiro** [esbmc/esbmc#580]
This issue is to remind us to develop a new option in ESBMC (e.g., `--undef-behaviour-check`) to check for undefined behaviour. In particular, this check would add an `assert(const >= 0)` or `assert(const <= type-bit-width)` statement for shift operations. It should be something similar to the `--overflow-check` (available in the `goto_check.cpp` file).

**Comment by mikhailramalho** [comment]
Also, `--undef-behaviour-check` would imply `--overflow-check` for signed integers.

**Comment by mikhailramalho** [comment]
CBMC already have a check shifts greater than the bit-width: https://github.com/diffblue/cbmc/blob/7b40362f8bca68bbe1c0c03ca502fd9925df898a/src/analyses/goto_check_c.cpp#L405

**Comment by fbrausse** [comment]
In #579 the test case https://github.com/esbmc/esbmc/blob/7a3c12f8354c5c7d853531334b65664282b1adc4/regression/cbmc/01_cbmc_BV_Arithmetic4/main.c was identified as exhibiting UB. The new check should confirm that.

**Comment by fbrausse** [comment]
#615 makes `regression/esbmc/github_159_postdecrement_fail` a KNOWNBUG: https://github.com/esbmc/esbmc/blob/6d7ba105174dda1b792d79b1fb0fcb366b6720ea/regression/esbmc/github_159_postdecrement_fail/main.c#L8-L12 ESBMC did produce a correct negative verdict before, but for the wrong reason: The expression `b--` moves `b` to point before the object `Q`, which is UB (in contrast to pointing to one element behind it). Due to another issue (#613) ESBMC wrongly dereferenced the decremented `b` instead of the original one. With #615 it no longer does and thus fails to detect the UB. The checker should detect it.

**Comment by lucasccordeiro** [comment]
@brcfarias: Can I ask you to work on this issue?

**Comment by mikhailramalho** [comment]
Isn't this an umbrella issue for several ub check? We only implemented the shift ones

**Comment by lucasccordeiro** [comment]
Yes, you're correct. GitHub has closed this issue automatically.
